### ISSUE TITLE
Consider synthesizing line separator and paragraph separator as space

### ISSUE THREAD

**Issue by tursunova** [harfbuzz/harfbuzz#4279]
According to the [a unicode FAQ / recommendation](http://www.unicode.org/faq/unsup_char.html#2) the line-separator character (0x2028) and the paragraph-separator character (0x2029) should be displayed as space (0x0020). 

We are [in the process of changing Chrome](https://chromium-review.googlesource.com/c/chromium/src/+/4581187) to align closer with other browsers in synthesizing the line-separator character (0x2028) and the paragraph-separator character (0x2029) characters as white space for Blink rendering. Chrome as of today renders line separator (0x2028) and paragraph separator (0x2029) as either a missing glyph or l-sep/p-sep symbol depending on font glyph coverage and fallback availability, compare [crbug.com/550275](http://crbug.com/550275) - which is perceived as inconvenient or unexpected by a high number of users (26 stars on the bug). 

Firefox and Safari display line and paragraph separators as zero width space. Firefox is considering rendering them as space, compare: https://bugzilla.mozilla.org/show_bug.cgi?id=1467395 - which is closer to the recommendation.

We [landed a WPT](https://github.com/web-platform-tests/wpt/pull/40364) test to ensure they get rendered as space without objections from other browsers.

As the behaviour is the prevalent and intended behaviour, we propose to make this synthesization a behaviour implemented at least as an optional behaviour in HarfBuzz, similar to existing white space synthetisation. We’re curious to hear the HarfBuzz community’s opinion.

**Comment by behdad** [comment]
I'm generally fine with this. Hopefully we can implement it without a performance hit. I don't know of any adverse impact.

@jfkthame any objections to doing this by default?

**Comment by behdad** [comment]
I'm generally fine with this. Hopefully we can implement it without a performance hit. I don't know of any adverse impact.

@jfkthame any objections to doing this by default?

**Comment by jfkthame** [comment]
I don't see a problem with doing this in harfbuzz.

(From a browser point of view, though, it's currently unclear to me whether this would be the correct result. In https://bugzilla.mozilla.org/show_bug.cgi?id=1467395#c5, the commenter argues they should result in a "hard" line-break, and the current behavior of browsers (whether rendering them as a space, zero-width, or whatever) is a bug. But I'm not aware of any recent discussion around this. Perhaps it needs to go to the CSS WG for clarification.)

**Comment by behdad** [comment]
This would come with a cost to implement though, since currently if the font has a glyph for a character we always use that. But that is not desirable here. I wonder if we should push this down to Chrome's implementation of font funcs?

**Comment by drott** [comment]
We have a change where we do this in the font-func override that we have in Chromium: https://chromium-review.googlesource.com/c/chromium/src/+/4581187/8/third_party/blink/renderer/platform/fonts/shaping/harfbuzz_face.cc - we were just thinking as this seems to be common behavior, it might architecturally be better suited to be in HarfBuzz. 

I agree that it's not completely clear if the characters should be allowed to control line breaking in HTML. This would indeed need a CSS WG clarification. CC @kojiishi for opinions on whether L-SEP should have the power to do that.

**Comment by behdad** [comment]
> it might architecturally be better suited to be in HarfBuzz.

I agree with that. But then shouldn't we do the same with CR/LF as well? Also, what happens with HTML whitespace folding? Just trying to understand if there's something more general we can / should do, or these two chars are the extent of it.

I think I can do it in a way that doesn't have a performance penalty.

**Comment by behdad** [comment]
> I think I can do it in a way that doesn't have a performance penalty.

If it was to synthesize them only when the font has no glyphs for them then I could implement without penalty. But if the goal is to replace them with space even if the font has inked glyphs for them, then I cannot think of a way to implement this without slowing down shaping all glyphs all the time.

Also waiting for answers to CR/LF and possibly other characters before I can do anything here.
